## Written by Courtney Smith 7-4-2023
## Goal of script: Generate dendrogram and heatmap from (subsetted) haplotypes and heatmap of (non-subsetted) regression results

library(dplyr)
library(ggplot2)
library("dendextend")
library("RColorBrewer")
library("corrplot")
library(ComplexHeatmap)

# Set parameters
nsnps=c(1000) #c(30,100,120)
blocks=1:3
pcut <- 1e-6
min_threshold=20000
ntraits=1
zcut = 4
extended = "N"

nsnp=1000

bnum=1

for (nsnp in nsnps){
  for (bnum in blocks){
    #### SUBSET DOWN TO CLUSTERS AND TRAITS TO FOCUS ON
    # set directories
    hla_dir="/home/ivm/haplo_analysis/"
    nsnp_dir=nsnp_dir=paste0(hla_dir,"snps",nsnp,"/")
    setwd(nsnp_dir)
    
    # load data
    c <- data.table::fread(paste0("hb",bnum,"_clusterstats.txt")) # cluster info, generated by hapclusterandplot.R
    m <- data.table::fread(paste0(nsnp_dir,"hb",bnum,"_mungedgenotypes01.txt")) # load in genptyoe info, generated by hapclusterandplot.R
    h1 <- data.table::fread(paste0("hb",bnum,"_regresults.txt")) %>% rename(trait=nimi)
    
    ## Get stats
    # how many unique haplotypes total
    length(unique(m$hap))
  
    # how many traits have at least one cluster w/ p < pcut of all clusters
    length(unique(filter(h1,p<pcut)$trait)) # (of the 269 tested)

    ## Calculate z-scores and add in dropped cluster
    h <- h1 %>% mutate(z=beta/se)
    beta <- tidyr::spread(h %>% select(trait,cluster,beta),cluster,beta)
    z <- tidyr::spread(h %>% select(trait,cluster,z),cluster,z)
    traits <- z$trait
    rownames(z) <- traits
    z <- z %>% select(-trait)
    # add in dropped cluster w/ rescale
    z <- z %>% mutate(dropped=0)
    # Function to rescale row values to have mean 0
    rescale_mean_zero <- function(row) {
      mean_zero <- row - mean(row)
      return(mean_zero)
    }
    # Apply the rescaling function to each row of the dataframe
    z_full <- t(apply(z, 1, rescale_mean_zero))
    rownames(z_full) <- traits
    
    # identify dropped cluster and rename last column added (dropped column) with it
    dropped <- filter(c,cluster_total==max(c$cluster_total))$Cluster
    numcol <- ncol(z_full)
    colnames(z_full) <- c(head(colnames(z_full),numcol-1),dropped)
    droppedextra <- setdiff(c$Cluster,colnames(z_full))
    
    # filter to a small number of traits of interest (have >= 1 clusters w/ abs(z)>zcut in the clusters to plot)
    z_filt <- z_full[(rowSums(abs(z_full)>zcut))>0,]
    
    # filter to a subset of clusters that have total doses > min_cutoff or at least 1 trait with abs(z)>zcut
    mckeep <- (c %>% filter(cluster_total>min_threshold))$Cluster
    mckeep <- mckeep[!mckeep %in% droppedextra]
    clustkeep <- unique(c(mckeep,colnames(z_full[,(colSums(abs(z_full)>zcut))>0])))
    z_filt <- as.data.frame(z_filt) %>% select(all_of(clustkeep))
    
    # stats
    dim(z_filt) # rows is traits, columns is clusters
    
    # save these results
    data.table::fwrite(z_filt, paste0(nsnp_dir,"hb",bnum,"_regdatatoplot.txt"), row.names = T, quote = F, sep = "\t")
    data.table::fwrite(as.data.frame(z_full), paste0(nsnp_dir,"hb",bnum,"_regdatafull.txt"), row.names = T, quote = F, sep = "\t")
    
    ### Add color assignments
    # Step 1: Generate colors
    colors = grDevices::colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]
    # Step 2: Convert to RGB
    rgb_values = col2rgb(colors)
    # Step 3: Calculate brightness
    brightness = 0.299*rgb_values[1,] + 0.587*rgb_values[2,] + 0.114*rgb_values[3,]
    # Step 4: Filter colors
    threshold = 150
    col_vector = colors[brightness < threshold]
    #col_vector = c("dodgerblue2", "#E31A1C", "green4","#6A3D9A", "#FF7F00","black", "gold1", "skyblue2", "#FB9A99", "palegreen2",  "#CAB2D6", "#FDBF6F", "gray70", "khaki2",  "maroon", "orchid1", "deeppink1", "blue1", "steelblue4",  "darkturquoise", "green1", "yellow4", "yellow3","darkorange4", "brown")
    colors <- data.frame(cluster=colnames(z_filt),
                         color= sample(col_vector,size=ncol(z_filt)))
    
    #### MAKE THE SUBSETDOWN DENDROGRAM AND SNP HEATMAP
    # select the n most common haplotypes from each cluster, only clusters to keep
    nkeep=40
    subset_df <- m %>% group_by(final_cluster) %>% arrange(-total) %>% slice(1:nkeep)
    subset_df <- subset_df %>% filter(final_cluster %in% clustkeep)
    df <- t(as.matrix(subset_df %>% ungroup %>% select(matches("chr6"))))
    l<- as.character(subset_df$final_cluster)
    C <- cluster_within_group(df,l)
    set.seed(123)
    Heatmap(t(df), cluster_rows = C, border=TRUE, row_split = length(unique(subset_df$final_cluster)),
            show_column_dend = FALSE,
            cluster_columns = FALSE,show_column_names=FALSE,
            show_row_names = TRUE,
            col=c("black","white"),show_heatmap_legend = FALSE,
            row_gap=unit(0.8,"mm"),
            row_title_gp = gpar(fontsize = 25),
            row_title_rot = 0)#,left_annotation = rowAnnotation(Cluster = l,show_legend=F))
    clustorder <- unique(l[order.dendrogram(C)]) # top to bottom
    
    #### PLOT THE REGRESSION RESULTS HEATMAP
    # Set parameters
    fdir="/home/ivm/haplo_analysis/figures/"
    
    # Reorder and name the clusters and subset down just to those to plot
    z <- z_filt
    #colnames(z) <- paste0("Cluster ",colnames(z))
    #clusternames <- unique(gsub(":.*","",sorted_order)) # sorted_order from the making of the subsetted dendrogram / snp 0s/1s heatmap
    #clusternames <- paste0("Cluster ",rev(clustorder))
    clusternames <- rev(clustorder)
    z <- as.data.frame(z) %>% select(clusternames) # reorder columns
    
    # Add cutoff values for plotting
    z[z<(-5)] <- -5
    z[z>(5)] <- 5
    
    # Cluster the traits axis of the heatmap (keep clusters in order of dendrogram not clustered here
    hc <- hclust(dist(z,method="euclidean"),method="ward.D")
    z <- z[hc$order,]
    z$traits <- rownames(z)
    z$traits <- factor(z$traits,levels=traits)
    melted_df <- reshape2::melt(z)
    
    # Add in updated trait names
    tn <- data.table::fread("/home/ivm/from_satu/general_files/new_plot_names.txt",header=T) %>%
      select(trait,Plot)
    
    melted_df <-melted_df %>% left_join(tn,by=c("traits"="trait"))
    
    # Plot the heatmap using ggplot2
    ggplot(melted_df, aes(x = Plot, y = variable, fill = value)) + # x=traits
      #geom_tile(color="black",size=1)+
      geom_tile()+
      #scale_fill_gradient2(low="blue",mid="white",high="red",midpoint=0)+
      scale_fill_gradient2(low="blue",mid="white",high="red",midpoint=0, limits = c(-5, 5))+
      theme_classic()+
      theme(text = element_text(size=25),
            axis.text.x = element_text(angle = -70, hjust = 0))+
      #labs(x="Traits",y="Clusters",fill="Z-score")
      labs(x="",y="",fill="Z-score")+
      scale_x_discrete(limits=unique(melted_df$Plot)) # z$traits
    ggsave(paste0(fdir,"snps",nsnp,"/",nsnp,"_b",bnum,"minthres_",min_threshold,"_traits",ntraits,"_cutoff5.png"))
    
    data.table::fwrite(melted_df,paste0(nsnp_dir,"hb",bnum,"_heatmapplotted.txt"), row.names = F, quote = F, sep = ",")
    
    cn <- data.frame(cluster=clusternames)
    data.table::fwrite(cn,paste0(nsnp_dir,"hb",bnum,"_heatmapclustorder.txt"), row.names = F, quote = F, sep = ",")
          if (extended=="N"){next}
    ### Same thing but for all clusters for all traits
    # Add cutoff values for plotting
    z <- z_full
    z[z<(-5)] <- -5
    z[z>(5)] <- 5
    
    # Cluster the traits axis of the heatmap (keep clusters in order of dendrogram not clustered here
    hc <- hclust(dist(z,method="euclidean"),method="ward.D")
    z <- z[hc$order,]
    melted_df_full <- reshape2::melt(z)
    ggplot(melted_df_full, aes(x = Var1, y = Var2, fill = value)) +
      #geom_tile(color="black",size=1)+
      geom_tile()+
      #scale_fill_gradient2(low="blue",mid="white",high="red",midpoint=0)+
      scale_fill_gradient2(low="blue",mid="white",high="red",midpoint=0, limits = c(-5, 5))+
      theme_classic()+
      theme(axis.text.x = element_text(angle = 70, hjust = 1))+
      #labs(x="Traits",y="Clusters",fill="Z-score")
      labs(x="",y="",fill="Z-score")
  }
}

### Going through stats
c <-  data.table::fread(paste0("hb",bnum,"_clusterstats.txt")) #%>% select(-Merged_Haplotype)
h <- data.table::fread(paste0("hb",bnum,"_haplostats.txt"))
c <- c %>% mutate(frac_twodose=cluster_twodose/cluster_unippl)
length(unique(m$hap)) # total number unqiue hap
dim(c) # number cluster total
dim(h %>% group_by(Cluster) %>% count() %>% filter(n>100))
dim(h %>% group_by(Cluster) %>% count() %>% filter(n>1000))
max(((h %>% group_by(Cluster) %>% count()))$n) # max num haplotypes in cluster
summary(c %>% select(cluster_total))
nrow((c %>% filter(cluster_total>10000))) # number clusters with > n total doses
nrow((c %>% filter(cluster_total>20000))) # number clusters with > n total doses
dim(z_filt) # traits by clusters plotted

############ SAVE SUPPLEMENTARY TABLES 3
library(dplyr)
nsnp=1000
hla_dir="/home/ivm/haplo_analysis/"
nsnp_dir=nsnp_dir=paste0(hla_dir,"snps",nsnp,"/")
setwd(nsnp_dir)

c1 <-  data.table::fread(paste0("hb",1,"_clusterstats.txt")) #%>% select(-Merged_Haplotype)
c2 <-  data.table::fread(paste0("hb",2,"_clusterstats.txt")) #%>% select(-Merged_Haplotype)
c3 <-  data.table::fread(paste0("hb",3,"_clusterstats.txt")) #%>% select(-Merged_Haplotype)
c <- bind_rows(c1 %>% mutate(block=1),c2 %>% mutate(block=2),c3 %>% mutate(block=3))
c <- c %>% mutate(frac_twodose=cluster_twodose/cluster_unippl)
c <- c %>% rename(haplotype_group=Cluster,num_onedose=cluster_onedose,
             num_twodose=cluster_twodose,num_uniquepeople=cluster_unippl,
             total_doses=cluster_total)
data.table::fwrite(c, paste0(nsnp_dir,"allhaplotype_group_stats.txt"), row.names = F, quote = F, sep = "\t")

h1 <- data.table::fread(paste0("hb",1,"_haplostats.txt"))
h2 <- data.table::fread(paste0("hb",2,"_haplostats.txt"))
h3 <- data.table::fread(paste0("hb",3,"_haplostats.txt"))
h <- bind_rows(h1 %>% mutate(block=1),h2 %>% mutate(block=2),h3 %>% mutate(block=3))
h <- h %>% mutate(frac_twodose=twodose/unippl)
h <- h %>% select(-dendroname,-color) %>% rename(haplotype_group=Cluster,
                                      num_onedose=onedose,
                                      num_twodose=twodose,
                                      num_uniquepeople=unippl,
                                            total_doses=total)
data.table::fwrite(h, paste0(nsnp_dir,"allhaplotype_stats.txt"), row.names = F, quote = F, sep = "\t")
data.table::fwrite(h %>% filter(total_doses>10), paste0(nsnp_dir,"allhaplotype_stats_morethan10.txt"), row.names = F, quote = F, sep = "\t")
